tracer/src/Datadog.Trace/AppSec/Security.cs

```diff
@@ -281,60 +278,112 @@ private void SetRemoteConfigCapabilites()
+            }
         }
 
         private void FeaturesProductConfigChanged(object sender, ProductConfigChangedEventArgs e)
         {
-            var features = e.GetDeserializedConfigurations<AsmFeatures>().FirstOrDefault();
-            if (features.TypedFile != null)
+            lock (_settings)
             {
-                _settings.Enabled = features.TypedFile.Asm.Enabled;
-                UpdateStatus(true);
-            }
+                var features = e.GetDeserializedConfigurations<AsmFeatures>().FirstOrDefault();
+                if (features.TypedFile != null)
+                {
+                    _settings.Enabled = features.TypedFile.Asm.Enabled;
+                    UpdateStatus(true);
+                }
 
-            e.Acknowledge(features.Name);
+                e.Acknowledge(features.Name);
+            }
         }
 
         private void AsmDataProductConfigChanged(object sender, ProductConfigChangedEventArgs e)
         {
-            if (!_enabled)
-            {
-                return;
-            }
+            if (!_enabled) { return; }
 
-            _asmDataConfigs ??= new Dictionary<string, Payload>();
-            var asmDataConfigs = e.GetDeserializedConfigurations<Payload>();
-            foreach (var asmDataConfig in asmDataConfigs)
+            lock (_settings)
             {
-                _asmDataConfigs[asmDataConfig.Name] = asmDataConfig.TypedFile;
-                e.Acknowledge(asmDataConfig.Name);
-            }
+                _asmDataConfigs ??= new Dictionary<string, RcmModels.AsmData.Payload>();
+                var asmDataConfigs = e.GetDeserializedConfigurations<RcmModels.AsmData.Payload>();
+                foreach (var asmDataConfig in asmDataConfigs)
+                {
+                    _asmDataConfigs[asmDataConfig.Name] = asmDataConfig.TypedFile;
+                    e.Acknowledge(asmDataConfig.Name);
+                }
 
-            var updated = UpdateRulesData();
-            foreach (var asmDataConfig in asmDataConfigs)
-            {
-                if (!updated)
+                var updated = UpdateRulesData();
+                foreach (var asmDataConfig in asmDataConfigs)
                 {
-                    e.Error(asmDataConfig.Name, "Waf could not update the rules");
+                    if (!updated)
+                    {
+                        e.Error(asmDataConfig.Name, "Waf could not update the rules");
+                    }
+                    else
+                    {
+                        e.Acknowledge(asmDataConfig.Name);
+                    }
                 }
-                else
+            }
+        }
+
+        private void AsmProductConfigChanged(object sender, ProductConfigChangedEventArgs e)
+        {
+            if (!_enabled) { return; }
+
+            lock (_settings)
+            {
+                _ruleStatus ??= new Dictionary<string, bool>(StringComparer.InvariantCultureIgnoreCase);
```

Comment:
Rather than locking while the dictionary is being built, wouldn't it be better to built the dictionary in reference held on the stack, then push reference to the field inside a lock (or atomic operation).

Reply:
True